extended dims cube

atolopko-czi opened the pull request:
For https://github.com/chanzuckerberg/cellxgene-census/issues/603

atolopko-czi committed:
convert to runnable module

atolopko-czi committed:
support logical cube dims as TileDB Dims and Attrs

atolopko-czi committed:
rm obsolete comment

atolopko-czi committed:
output array names

atolopko-czi commented on memento/cell_census_summary_cube.py:
```diff
@@ -30,53 +30,49 @@
     "donor_id",
     "disease",
     "sex"
 ]
```
@mincheoly The above split of columns between "dims" and "attrs" allows us to select the columns that TileDB will index (in the sense of a relational database), allowing for efficient filtering on these columns. What is the access pattern for the cube? If there are 7 columns, will filtering always be performed on all columns? Or is it arbitrary?

atolopko-czi committed:
fix logging pct

atolopko-czi committed:
add profiling for pass 2 (single-process)

atolopko-czi commented:
@mincheoly This code is currently the performance hotspot:
```
gene_rvs = gen.multinomial(n_obs, counts / counts.sum(), size=num_boot).T
```
With the extra dimensions, this is being called considerably more times. The new cube will have ~102970 unique dimension tuples vs the original cube's ~4047, which means 25x more `multinomial` calls.

Is there a faster implementation we can find?

Can we possibly decrease `num_boot`, as this helps it run faster?

mincheoly commented:
i think reducing num_boot to ~1000 (maybe even 500) should be fine (and we can compare later on a smaller cube?)

I'm surprised that's the bottleneck and not getting the unique counts. i remember testing a few different ways to generate the multinomial and they didnt seem hugely different - but maybe things have changed since (like https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.multinomial.html)

Would reducing the num_boot get it to a point where its runnable?

atolopko-czi commented:
> i think reducing num_boot to ~1000 (maybe even 500) should be fine (and we can compare later on a smaller cube?)

Great! I'll start with 500 and see that affects performance.

> Would reducing the num_boot get it to a point where its runnable?

It should at least make that one call 10x faster, and since it's the performance hotspot, that may be enough. Stay tuned!

atolopko-czi committed:
reinstate parallelism

atolopko-czi committed:
update local testing config (commented-out)

atolopko-czi committed:
update local testing config (commented-out)

atolopko-czi committed:
optimizations

- use numba jit for multinomial generation
- eliminate redundant calculation

Commit:
set multinomial size to 500 for performance

atolopko-czi committed:
improve profiling support

- profile pass 2 comprehensively
- parameterize profiling mode (and test mode) via env vars

atolopko-czi committed:
Merge branch 'atol/extended-dims-cube' of github.com:mincheoly/memento-cxg into atol/extended-dims-cube

atolopko-czi committed:
cleanup

atolopko-czi commented:
@mincheoly (cc: @pablo-gar) Since we can generate the cube in ~1 day, perhaps that is sufficient for now? I figure that as long as you can get the the larger dimension cube generated once, you can experiment with it. In which case, I'll pause on doing further optimization work. And if we get to the point of needing to generate this cube on a regular basis (e.g. for every Census data release), I will put more effort into optimizing.

I think the code changes are worth merging to main now, as is, but I'll wait for your approval on this PR before merging (no rush). Thanks!
